singleflight/example_test.go

```diff
@@ -0,0 +1,37 @@
+package singleflight
+
+import (
+	"fmt"
+	"sync"
+	"time"
+)
+
+func Example() {
+	// Spin up 5 goroutines, each of which calls a function that increments
+	// a counter. At the end, the counter will only increment once and the
+	// counter value (1) will be returned to all goroutines.
+	counter := 0
+	var g Group
+	var wg sync.WaitGroup
+	var results [5]int
+	c := make(chan bool)
+	for i := 0; i < 5; i++ {
+		wg.Add(1)
+		go func(i int) {
+			ctrval, _ := g.Do("key", func() (interface{}, error) {
+				counter++
+				<-c // make sure every goroutine has started before returning
+				return counter, nil
+			})
+			results[i] = ctrval.(int)
+			wg.Done()
+		}(i)
+	}
+	// Sleep so all of the goroutines start before we send a message on the
+	// channel
+	time.Sleep(10 * time.Millisecond)
```

Comment:
I'm never a fan of sleeps in tests. Use a channel or something to wait for everybody to get in there instead.